pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,353 @@
+				"glusterbkpservers": strings.Join(bkpServers, ":"),
+			},
+		},
+	}
+
+	glog.V(4).Infof("CSI Volume response: %+v", resp)
+	return resp, nil
+}
+
+func (cs *ControllerServer) checkExistingVolumes(volumeListResp *api.VolumeListResp, volumeName string, volSizeBytes int) (string, []string, error) {
+
+	var tspServers []string
+	var mountServer string
+	var err error
+
+	if volumeListResp == nil {
+		glog.Errorf("Provided volume list response is nil")
+		return "", nil, fmt.Errorf("volume list response is nil")
+	}
+
+	for _, vol := range *volumeListResp {
+		if vol.Name == volumeName {
+			// Do the owner validation
+			if glusterAnnVal, found := vol.Metadata[glusterDescAnn]; found {
+				if glusterAnnVal != glusterDescAnnValue {
+					return "", nil, status.Errorf(codes.Internal, "volume %s (%s) is not owned by Gluster CSI driver",
+						vol.Name, vol.Metadata)
+				}
+			} else {
+				return "", nil, status.Errorf(codes.Internal, "volume %s (%s) is not owned by Gluster CSI driver",
+					vol.Name, vol.Metadata)
+			}
+			vsResp, e := cs.client.VolumeStatus(vol.ID.String())
+			if e != nil {
+				return "", nil, status.Errorf(codes.Internal, "failed to get volume status %s", e.Error())
+			}
+			if int(vsResp.Size.Capacity) != volSizeBytes {
+				return "", nil, status.Error(codes.AlreadyExists, fmt.Sprintf("volume already exits with different size: %d", vsResp.Size.Capacity))
+			}
+
+			glog.Info("Requested volume (%s) already exists in the storage pool", vol.Name)
+			mountServer, tspServers, err = cs.getClusterNodes()
+
+			if err != nil {
+				return "", nil, status.Error(codes.Internal, fmt.Sprintf("error in fetching backup/peer server details %s", err.Error()))
+			}
+
+		}
+	}
+
+	return mountServer, tspServers, nil
+
+}
+func (cs *ControllerServer) getClusterNodes() (string, []string, error) {
+
+	peers, err := cs.client.Peers()
+	if err != nil {
+		return "", nil, err
+	}
+	glusterServer := ""
+	bkpservers := []string{}
+
+	for i, p := range peers {
+		if i == 0 {
+			for _, a := range p.PeerAddresses {
+				ip := strings.Split(a, ":")
+				glusterServer = ip[0]
+			}
+
+			continue
+		}
+		for _, a := range p.PeerAddresses {
+			ip := strings.Split(a, ":")
+			bkpservers = append(bkpservers, ip[0])
+		}
+
+	}
+	glog.V(2).Infof("Gluster server and Backup servers [%+v,%+v]", glusterServer, bkpservers)
+
+	return glusterServer, bkpservers, err
+}
+
+// DeleteVolume deletes the given volume. The function is idempotent.
+func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) (*csi.DeleteVolumeResponse, error) {
+	// TODO; return success if volume not found
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Volume delete request is nil")
+	}
+
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "volume ID is nil")
+	}
+	glog.V(2).Infof("Deleting volume with ID: %v", req.VolumeId)
+
+	err := cs.client.VolumeStop(req.VolumeId)
```

Comment:
Is it CSI plugin's responsibility to control the status of the volume?
If yes, why do we not start volume during volume create?

Reply:
we are starting the volume during the volume creation, as part of volume delete we are stopping the volume